src/database/arangodb/aql-generator.ts

```diff
@@ -1623,54 +2344,131 @@ function getRelationTraversalFragment({
+    // This does not apply to the last FOR though
+    // - if it's not a list, this is correct: we would want the NULL values there
+    //   (e.g. deliveries.order should have as many NULLs as there are deliveries without an order)
+    //   dangling edges are already filtered out above in the LET (needsNullableVar)
+    // - if it's a list, we shouldn't have any NULLs because we don't have "relation to NULL"s
+    //   in case of dangling edges however (i.e. edge without corresponding document)
+    //   we could generate NULLS. We generally ignore those in cruddl. -> filter them out
+    //   preserveNullValues is handled above (the dangling edge filter is in the LET there)
+    if (segments[segments.length - 1].isListSegment) {
         // using $var._key != null instead of $var != null because the latter prevents ArangoDB
         // from applying the reduce-extraction-to-projection optimization
         forFragments.push(aql`FILTER ${currentObjectFrag}._key != null`);
     }
 
-    const returnFrag = mapFrag ? mapFrag(currentObjectFrag) : currentObjectFrag;
-    const returnList = lastSegment.resultIsList || sourceIsList || alwaysProduceList;
     // make sure we don't return a list with one element
-    return aqlExt[returnList ? 'parenthesizeList' : 'parenthesizeObject'](
-        sourceIsList ? aql`FOR ${sourceEntityVar} IN ${sourceFrag}` : aql``,
-        ...forFragments,
-        aql`RETURN ${returnFrag}`,
-    );
+    return aql.lines(...forFragments);
 }
 
-function getFlattenFrag(listFrag: AQLFragment, depth: number) {
-    if (depth <= 0) {
-        return listFrag;
-    }
-    if (depth === 1) {
-        return aql`${listFrag}[**]`;
-    }
-    return aql`FLATTEN(${listFrag}, ${aql.integer(depth)})`;
+interface FieldTraversalFragment {
+    readonly segments: ReadonlyArray<FieldSegment>;
+    readonly sourceFrag: AQLFragment;
+    /**
+     * Note: cannot contain any subqueries (not supported by ArangoDB)
+     */
+    readonly mapFrag?: (itemFrag: AQLFragment) => AQLFragment;
+    readonly filterFrag?: (itemFrag: AQLFragment) => AQLFragment;
+    readonly skip?: number;
+    readonly maxCount?: number;
 }
 
-function getFieldTraversalFragmentWithoutFlattening(
-    segments: ReadonlyArray<FieldSegment>,
-    sourceFrag: AQLFragment,
-) {
+function getFieldTraversalFragment({
+    segments,
+    sourceFrag,
+    mapFrag,
+    filterFrag,
+    skip = 0,
+    maxCount,
+}: FieldTraversalFragment) {
     if (!segments.length) {
         return sourceFrag;
     }
+    if (
+        (mapFrag || filterFrag || maxCount !== undefined || skip > 0) &&
+        !segments.some((s) => s.isListSegment)
+    ) {
+        // if this is needed, it needs to happen outside the field traversal
+        // (it can just use the result of the field traversal as loop "variable")
+        throw new Error(
+            `Cannot have map, filter or limit on field traversal without list segments`,
+        );
```

Comment:
Changed the comment to make it easier understand from my view. Feel free to keep the original comment if you think it is better or mine is factually not 100% correct.

```typescript
// map-, filter- or limit-fragments on field traversal are only applicable in an array expansion which in turn can only be applied to lists
// otherwise the filter needs to be applied on the result of the field traversal, i.e. by using the result of the field traversal as the loop variable
throw new Error(
   `Cannot have map, filter or limit on field traversal without list segments`,
);
```